src/AggregateFunctions/AggregateFunctionFlameGraph.cpp

```diff
@@ -0,0 +1,574 @@
+
+                nodes.pop_back();
+                allocated_total.pop_back();
+                allocated_self.pop_back();
+
+                /// We don't have root's frame so framers are empty in the end.
+                if (!frames.empty())
+                    frames.pop_back();
+
+                continue;
+            }
+
+            TreeNode * current = nodes.back()->child;
+            nodes.back() = nodes.back()->next;
+
+            bool enough_bytes = current->allocated >= min_bytes;
+            bool enough_depth = max_depth == 0 || nodes.size() < max_depth;
+
+            if (enough_bytes)
+            {
+                frames.push_back(current->ptr);
+                allocated_self.back() -= current->allocated;
+
+                if (enough_depth)
+                {
+                    allocated_total.push_back(current->allocated);
+                    allocated_self.push_back(current->allocated);
+                    nodes.push_back(current->children);
+                }
+                else
+                {
+                    traces.push_back({frames, current->allocated, current->allocated});
+                    frames.pop_back();
+                }
+            }
+        }
+
+        return traces;
+    }
+};
+
+static void insertData(DB::PaddedPODArray<UInt8> & chars, DB::PaddedPODArray<UInt64> & offsets, const char * pos, size_t length)
+{
+    const size_t old_size = chars.size();
+    const size_t new_size = old_size + length + 1;
+
+    chars.resize(new_size);
+    if (length)
+        memcpy(chars.data() + old_size, pos, length);
+    chars[old_size + length] = 0;
+    offsets.push_back(new_size);
+}
+
+/// Split str by line feed and write as separate row to ColumnString.
+static void fillColumn(DB::PaddedPODArray<UInt8> & chars, DB::PaddedPODArray<UInt64> & offsets, const std::string & str)
+{
+    size_t start = 0;
+    size_t end = 0;
+    size_t size = str.size();
+
+    while (end < size)
+    {
+        if (str[end] == '\n')
+        {
+            insertData(chars, offsets, str.data() + start, end - start);
+            start = end + 1;
+        }
+
+        ++end;
+    }
+
+    if (start < end)
+        insertData(chars, offsets, str.data() + start, end - start);
+}
+
+void dumpFlameGraph(
+    const AggregateFunctionFlameGraphTree::Traces & traces,
+    DB::PaddedPODArray<UInt8> & chars,
+    DB::PaddedPODArray<UInt64> & offsets)
+{
+    DB::WriteBufferFromOwnString out;
+
+    std::unordered_map<uintptr_t, size_t> mapping;
+
+#if defined(__ELF__) && !defined(OS_FREEBSD)
+    auto symbol_index_ptr = DB::SymbolIndex::instance();
+    const DB::SymbolIndex & symbol_index = *symbol_index_ptr;
+#endif
+
+    for (const auto & trace : traces)
+    {
+        if (trace.allocated_self == 0)
+            continue;
+
+        for (size_t i = 0; i < trace.frames.size(); ++i)
+        {
+            if (i)
+                out << ";";
```

Comment:
There are two anti-patterns here:

1. Returning serialized data as JSON in a String.
We should return structured data and make it possible to serialize in any possible way outside of this aggregate function.
The previous incident of this offense was successfully resolved: https://github.com/ClickHouse/ClickHouse/pull/29934

2. Symbolization inside the aggregate function. It is slow and sometimes user may need a different kind of symbolization or no symbolization at all. By the way, it is also insecure unless `allow_introspection_functions` is checked.

Maybe make the aggregate function structured and generic and implement all the decorations with SQL UDF (can be preloaded if needed)?

Reply:
Note: we also have this offense in EXPLAIN, but less criminal.

Reply:
I can implement a function in a way that it returns a stacktrace  + sizes, the result would be a `Array(Tuple(Array(UInt64), UInt64))`. It can be used like
```
select arrayStringConcat(arrayMap(trace -> demangle(addressToSymbol(trace)), s.0), ';') || ' ' || toString(s.1) from (select arrarJoin(flameGraph(trace, size, ptr)) as s from system.trace_log where ...
```

On the other case, I think that using just
```
select arrarJoin(flameGraph(trace, size, ptr)) from system.trace_log
```
is much more user-friendly.

Reply:
It is not a JSON, it is a custom format from [here](https://github.com/brendangregg/FlameGraph/blob/master/flamegraph.pl#L24).